rasa/core/policies/two_stage_fallback.py

```diff
@@ -172,11 +175,17 @@ def predict_action_probabilities(
         return result
 
     def _is_user_input_expected(self, tracker: DialogueStateTracker) -> bool:
-        return tracker.latest_action_name in [
+        action_requires_input = tracker.latest_action_name in [
```

Comment:
shouldn't the latest action be an `ACTION_LISTEN` which wouldn't trigger this? If you change the test down here as I suggested, then it also passes without this change 🙈

Reply:
in case there is no `ACTION_LISTEN` in reality, then I think the bug is more the missing `action_listen` after the `self.fallback_action_name`, isn't it?

Reply:
which in turn is probably caused by the dispatched `utter_` in the fallback action, right?

Reply:
You're right, there is no `ACTION_LISTEN` in the tracker:
```
{
  "sender_id": "default",
  "slots": {
    "budget": null,
    "business_email": null,
    "can_use_spacy": null,
    "company": null,
    "current_api": null,
    "email": null,
    "entity_extractor": null,
    "feedback_message": null,
    "feedback_value": "negative",
    "job_function": null,
    "language": null,
    "name": null,
    "nlu_part": null,
    "onboarding": null,
    "person_name": null,
    "problem_description": null,
    "product": null,
    "requested_slot": null,
    "shown_privacy": null,
    "step": null,
    "suggestion": null,
    "unknown_nlu_part": null,
    "unknown_product": null,
    "use_case": null
  },
  "latest_message": {
    "text": "\/restart",
    "intent": {
      "name": "restart",
      "confidence": 1.0
    },
    "intent_ranking": [
      {
        "name": "restart",
        "confidence": 1.0
      }
    ],
    "entities": []
  },
  "latest_event_time": 1579872158.2922680378,
  "followup_action": null,
  "paused": true,
  "events": [
    {
      "event": "action",
      "timestamp": 1579872150.3798677921,
      "name": "action_session_start",
      "policy": null,
      "confidence": null
    },
    {
      "event": "session_started",
      "timestamp": 1579872150.3798770905
    },
    {
      "event": "action",
      "timestamp": 1579872150.3799209595,
      "name": "action_listen",
      "policy": null,
      "confidence": null
    },
    {
      "event": "user",
      "timestamp": 1579872150.3809549809,
      "text": "\/greet@0.75",
      "parse_data": {
        "text": "\/greet@0.75",
        "intent": {
          "name": "greet",
          "confidence": 0.75
        },
        "intent_ranking": [
          {
            "name": "greet",
            "confidence": 0.75
          }
        ],
        "entities": []
      },
      "input_channel": "cmdline",
      "message_id": "ef213f686b9d42a5894d45edff55fad4",
      "metadata": {}
    },
    {
      "event": "action",
      "timestamp": 1579872150.630562067,
      "name": "action_default_ask_affirmation",
      "policy": "policy_2_TwoStageFallbackPolicy",
      "confidence": 1.0
    },
    {
      "event": "bot",
      "timestamp": 1579872150.6305680275,
      "text": "Sorry, I'm not sure I've understood you correctly 🤔 Do you mean...",
      "data": {
        "elements": null,
        "quick_replies": null,
        "buttons": [
          {
            "title": "Say \"Hi!\" ",
            "payload": "\/greet{}"
          },
          {
            "title": "Something else",
            "payload": "I am asking you an out of scope question"
          }
        ],
        "attachment": null,
        "image": null,
        "custom": null
      },
      "metadata": {}
    },
    {
      "event": "action",
      "timestamp": 1579872150.6371781826,
      "name": "action_listen",
      "policy": "policy_2_TwoStageFallbackPolicy",
      "confidence": 1.0
    },
    {
      "event": "user",
      "timestamp": 1579872156.2601130009,
      "text": "\/greet@0.75",
      "parse_data": {
        "text": "\/greet@0.75",
        "intent": {
          "name": "greet",
          "confidence": 0.75
        },
        "intent_ranking": [
          {
            "name": "greet",
            "confidence": 0.75
          }
        ],
        "entities": []
      },
      "input_channel": "cmdline",
      "message_id": "3497321f520e437e9c06e2497503afa6",
      "metadata": {}
    },
    {
      "event": "action",
      "timestamp": 1579872156.2789099216,
      "name": "action_default_fallback",
      "policy": "policy_2_TwoStageFallbackPolicy",
      "confidence": 1.0
    },
    {
      "event": "bot",
      "timestamp": 1579872156.2789177895,
      "text": "Click the button below if you want to start over.",
      "data": {
        "elements": null,
        "quick_replies": null,
        "buttons": [
          {
            "payload": "\/restart",
            "title": "Restart"
          }
        ],
        "attachment": null,
        "image": null,
        "custom": null
      },
      "metadata": {}
    },
    {
      "event": "slot",
      "timestamp": 1579872156.278922081,
      "name": "feedback_value",
      "value": "negative"
    },
    {
      "event": "pause",
      "timestamp": 1579872156.2789261341
    },
    {
      "event": "user",
      "timestamp": 1579872158.2857079506,
      "text": "\/restart",
      "parse_data": {
        "text": "\/restart",
        "intent": {
          "name": "restart",
          "confidence": 1.0
        },
        "intent_ranking": [
          {
            "name": "restart",
            "confidence": 1.0
          }
        ],
        "entities": []
      },
      "input_channel": "cmdline",
      "message_id": "7f2b80fca755494e95ee239e8d65fb44",
      "metadata": {}
    },
    {
      "event": "action",
      "timestamp": 1579872158.2922680378,
      "name": "action_listen",
      "policy": "policy_2_TwoStageFallbackPolicy",
      "confidence": 1.0
    }
  ],
  "latest_input_channel": "cmdline",
  "active_form": {},
  "latest_action_name": "action_listen"
}
```

Reply:
It actually looks a lot like the issue addressed here: https://github.com/RasaHQ/rasa/issues/4107 but that was already fixed 🤔

Reply:
^ ok no, that's not the issue, the mapping policy successfully predicts `ACTION_RESTART`, it's just getting overridden.

Reply:
what is the difference between `pausing` a conversation and executing `ACTION_LISTEN`?

Reply:
Good question :-D Pausing means that the bot doesn't respond anymore. In my opinion there should be an action listen after the SlotSetEvent. What happens if you put a `FollowupAction(name="action_listen")` in the returned events in the actions in sara instead of the ConversationPausedEvent?

Reply:
Yup, that fixes it too. Should any bot output that waits for input end with `action_listen`? If so, then this is a bug in the user's (incl. sara's) action code, not rasa, right?

Reply:
Yeah, that's the question. From a user's perspective it's not really great because it's not intuitive. If you don't do the followup event and no conversation paused, how does the tracker look like?

I think a solution could also be to do: 

```python
    def _is_user_input_expected(self, tracker: DialogueStateTracker) -> bool:
        return tracker.latest_action_name in [
            ACTION_DEFAULT_ASK_AFFIRMATION_NAME,
            ACTION_DEFAULT_ASK_REPHRASE_NAME,
            self.fallback_action_name,
        ] or tracker.is_paused()
```

Reply:
In that case the tracker has `action_listen`:
```
    {
      "event": "bot",
      "timestamp": 1579876926.796528101,
      "text": "Click the button below if you want to start over.",
      "data": {
        "elements": null,
        "quick_replies": null,
        "buttons": [
          {
            "payload": "\/restart",
            "title": "Restart"
          }
        ],
        "attachment": null,
        "image": null,
        "custom": null
      },
      "metadata": {}
    },
    {
      "event": "slot",
      "timestamp": 1579876926.796530962,
      "name": "feedback_value",
      "value": "negative"
    },
    {
      "event": "action",
      "timestamp": 1579876926.8050029278,
      "name": "action_listen",
      "policy": "policy_2_TwoStageFallbackPolicy",
      "confidence": 1.0
    }
```

Reply:
We actually need to exclude `tracker.is_paused()` right? Otherwise it confirms the buggy behaviour of not restarting if the last action was `action_default_fallback`

Reply:
Exclude it `tracker.is_paused() ` from where?

Reply:
Sorry that was ambiguous - I mean in your solution instead of `or` shouldn't it be `and not`? 
```
def _is_user_input_expected(self, tracker: DialogueStateTracker) -> bool:
    return tracker.latest_action_name in [
            ACTION_DEFAULT_ASK_AFFIRMATION_NAME,
            ACTION_DEFAULT_ASK_REPHRASE_NAME,
            self.fallback_action_name,
        ] and not tracker.is_paused()
```
That way if the tracker is paused, `_is_user_input_expected` will return `False` and therefore `action_listen` won't be predicted (at least not with confidence 1.0)

Reply:
Yep, you're right 👍 So in that case the fallback policy should just return `self.fallback_scores(domain, self.core_threshold)`

Reply:
Added `ConversationPaused` to the events returned by `action_default_fallback`; and including that check - restarting then works, but if the user chooses not to restart and to provide their own message, since the tracker is paused the message is not handled, so that doesn't quite do it either.

Reply:
Actually, I'm not sure what the expected behaviour is if a user does not restart and chooses to type a message. atm it just loops back to `action_default_fallback`

Reply:
But I think that's the expected behavior, no? We are like "We can't repair the current situation - please restart". And if they don't restart and continue we just tell them "Hey, really: We can't do anything - please restart"

Reply:
Ok, in that case the pausing works. It does mean that anyone using this approach in their `action_default_fallback` must explicitly include either another action  or `ConversationPaused` in the events returned. If the `action_default_fallback` in rasa-demo is the standard way to do a restart fallback action then that makes sense.

Reply:
> must explicitly include either another action

Mhm, it could also return an empty list `[]` of events and then it would expect a `usermessage` right?

Reply:
If it returns an empty list o (or `[SlotSet]`, like in rasa-demo), then the `latest_action_name` will still be `action_default_fallback`, triggering `action_listen` instead of `restart`.

Reply:
triggering an `action_listen` by default is fine, right? Just shouldn't be the case if the conversation was explicitly paused.

Reply:
I might be missing something - can the conversation be paused while returning an empty list of events? I had been pausing it by returning `[ConversationPaused]`.

Reply:
>  can the conversation be paused while returning an empty list of events?

No. When returning an empty list, the conversation should just go on (although the bot might be completely off tracks)

Reply:
Ok. Then the problem in this case:
> it could also return an empty list `[]` of events and then it would expect a `usermessage` right?

> triggering an `action_listen` by default is fine

is that the user  has already given a message i.e. `/restart`, so if the bot does `action_listen` it doesn't restart. If the list of events returned contains any action besides `action_default_fallback`, it will restart (this without any of the changes in this pull request). 

This is why I originally was checking for user input already given; but we could also tell users that they should always return `action_listen` after asking for user input in a custom action. That would solve it without any code changes.

Reply:
Mhm, this is the default implementation of `action_default_fallback`: `https://github.com/RasaHQ/rasa/blob/9f74e945611de405e47dee9344057063e8ec087b/rasa/core/actions/action.py#L411` . But you're right that's probably not exactly user friendly.

Another thought I just had: Is that still the problem with rasa 1.6 ? There should be session restart after `restart` imo

Sorry, for this long discussion 🙈

Reply:
It's all good!  good  for me to learn to think through implications. 

> Is that still the problem with rasa 1.6 ?
I branched off of 1.6.x, so yes

Reply:
It's just in Sara yes, because we decided that just reverting the latest user message is just gonna leave a very messy state and we should rather start from scratch.

Ok, then let's check if there was user message after the latest action. I think we should be able to just compare timestamps, right? There is a helpful function called `DialogueStateTracker.get_last_event_for(...)` which you could use instead of hardcoding indexes

Reply:
updated - sorry for the delay

Reply:
@wochinge is this good now?